scripts/fetch_articles.py

```diff
@@ -1,181 +1,313 @@
+            for hit in epmc_hits:
+                # ignore preprints determined by two fields pubType and source
+                if 'pubType' in hit and hit['pubType'] == 'preprint':
+                    continue
+                if 'source' in hit and hit['source'] == 'PPR':
+                    continue
+                # determine the article id which will be used in ES, PMC id is preferred, because
+                # 1) it has PMC prefix rather than a string of digits
+                # 2) PMC guarantees open access, more likely to have dataset accession linked
+                article_id = determine_article_id(hit)
+                if len(article_id) == 0:
+                    logger.error(f"Study {dataset_id} has related article without Identifier")
+                    continue
+                # new article
+                if article_id not in article_details:
+                    es_article = dict()
+                    es_article = parse_field(es_article, hit, 'pmcId', 'pmcid')
+                    es_article = parse_field(es_article, hit, 'pubmedId', 'pmid')
+                    es_article = parse_field(es_article, hit, 'doi', 'doi')
+                    es_article = parse_field(es_article, hit, 'title', 'title')
+                    es_article = parse_field(es_article, hit, 'authorString', 'authorString')
+                    es_article = parse_field(es_article, hit, 'journal', 'journalTitle')
+                    es_article = parse_field(es_article, hit, 'issue', 'issue')
+                    es_article = parse_field(es_article, hit, 'volume', 'journalVolume')
+                    es_article = parse_field(es_article, hit, 'year', 'pubYear')
+                    es_article = parse_field(es_article, hit, 'pages', 'pageInfo')
+                    es_article = parse_field(es_article, hit, 'isOpenAccess', 'isOpenAccess')
+                    article_details[article_id] = es_article
+                    article_basic_info = dict()
+                    # the article information displayed in other entities
+                    article_basic_info['articleId'] = article_id
+                    article_basic_info = parse_field(article_basic_info, hit, 'title', 'title')
+                    article_basic_info = parse_field(article_basic_info, hit, 'journal', 'journalTitle')
+                    article_basic_info = parse_field(article_basic_info, hit, 'year', 'pubYear')
+                    article_basics[article_id] = article_basic_info
+
+                article_datasets.setdefault(article_id, set())
+                article_datasets[article_id].add(dataset_id)
+                article_for_datasets.setdefault(dataset_id, set())
+                article_for_datasets[dataset_id].add(article_id)
+
+    logger.info(f'Retrieved {len(article_details.keys())} articles from all datasets')
+
+    es = Elasticsearch(hosts)
+    # deal with article index
+    for article_id in article_details.keys():
```

Comment:
*.keys() is not required here, as by default python provide iteration over keys for dicts.

Reply:
initially thought more understandable for other language users
now removed in 712246d